### DISCUSSION TITLE
Init Resource in one system and immediately use in another

### DISCUSSION THREAD

**Discussion by romamik** [bevyengine/bevy#9763] [Q&A] [answered]
Here is the code. I understand why it does not work: when I call `commands.insert_resource` the resource is only scheduled to be inserted on command execution and this will happen only after all the systems have run. But how do I achieve the desired result?
```
use bevy::prelude::*;

#[derive(Resource)]
struct MyResource {
    value: i32,
}

#[derive(States, Debug, Hash, PartialEq, Eq, Clone, Default)]
enum GameState {
    #[default]
    Loading,
    Running,
}

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_state::<GameState>()
        .add_systems(Update, load_state.run_if(in_state(GameState::Loading)))
        .add_systems(OnEnter(GameState::Running), create_resource)
        .add_systems(
            OnEnter(GameState::Running),
            use_resource.after(create_resource),
        )
        .run();
}

fn load_state(mut next_state: ResMut<NextState<GameState>>) {
    println!("Loading state");
    next_state.set(GameState::Running);
}

fn create_resource(mut commands: Commands) {
    println!("Creating resource");
    commands.insert_resource(MyResource { value: 42 });
}

fn use_resource(my_resource: Res<MyResource>) {
    println!("Used resource value: {}", my_resource.value);
}
```

**Comment by alice-i-cecile** [comment, marked as answer]
Commands can be manually scheduled by inserting a copy of the [`apply_deferred`](https://docs.rs/bevy/latest/bevy/ecs/prelude/fn.apply_deferred.html) system between the system where you produce the command and where you need it to have been ready.

However, that's not your only good option here! You could also:

- use the `FromWorld` trait to perform more complex resource initialization
- use an exclusive system to directly insert the resource into the world, skipping the indirection of commands completely
